### PR TITLE
woodpecker-{agent,cli,server}: init at 0.15.3

### LINKED ISSUE #178332: Woodpecker CI server, runner, and CLI
**Project description**

[Woodpecker CI](https://woodpecker-ci.org) is a simple CI engine, [forked from Drone CI after their license change](https://woodpecker-ci.org/faq#why-is-woodpecker-a-fork-of-drone-version-08)

**Metadata**

* homepage URL: https://woodpecker-ci.org/
* source URL: https://github.com/woodpecker-ci/woodpecker
* license: apache 2.0
* platforms: all (since it's written in Go), but the docker backend is linux-specific

### PR DESCRIPTION
**Pull request by ambroisie**
###### Description of changes

Supersedes #169470.
Closes #178332.

Some notes :

* [x] We probably want to prefix the different binaries with `woodpecker-`, instead of just `cli`, `agent`, and `server`.
* [x] Address/remove the FIXME about `musl`.
* [x] Perhaps add an update script.

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by wyndon** [comment]
Do you plan to add a module?

**Comment by ambroisie** [comment]
Not sure why ofborg is unhappy about (what seems to be) the `mkYarnPackage` derivation.

**Comment by 06kellyjac** [comment]
they all just copy in package.json, its a simple wget or curl in the update script. Again I'll see if I can tweak yarn2nix to stop the IFD issues when I get some time

---

I made some changes https://github.com/NixOS/nixpkgs/commit/e243a8c9ec4043cb436ed2d570b6fb975cb9d47c

* adds prefixes to bins
* removes the need for libexec/xyz etc by just keeping the dist folder
  * before: `λ du -h -d 0 ./result/` -> `359M`
  * after: `λ du -h -d 0 ./result/` -> 664K
* seperate the frontend to its own input, this is important so people can fix it if it breaks like jellyfin and all other node stuff did recently https://github.com/NixOS/nixpkgs/issues/176288
* drop the extra static stuff, it seems to cause more issues than it solves, we don't need to exactly copy the upstream builds
  * LMK if this resolves the glibc musl stuff, not sure what the actual issue was in order to test it


server started and ui shows fine but that's all I did:

`WOODPECKER_GITHUB=true ./result/bin/woodpecker-server --server-host http://192.168.0.111`

image

**Comment by ambroisie** [comment]
@06kellyjac I think we should only keep the front-end as a `passthru` attribute, I don't think it makes much sense to have it as an actual package in nixpkgs but it would still allow overriding thusly (approximately? not tested):

```nix
self: super:
{
  woodpecker-server = super.woodpecker-server.override {
    woodpecker-frontend = super.woodpecker-server.passthru.woodpecker-frontend.override {
      # Override here
    };
  };
}
```

File: pkgs/development/tools/continuous-integration/woodpecker/frontend.nix [hunk 1]
```diff
@@ -0,0 +1,37 @@
+{ lib, fetchFromGitHub, mkYarnPackage }:
+let
+  inherit (import ./common.nix { inherit lib fetchFromGitHub; })
+    meta
+    version
+    src
```
Comment (06kellyjac) [inline pkgs/development/tools/continuous-integration/woodpecker/frontend.nix:6]:
```suggestion
    src
    yarnSha256
```

File: pkgs/development/tools/continuous-integration/woodpecker/common.nix [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+rec {
+  version = "0.15.3";
+
+  src = fetchFromGitHub {
+    owner = "woodpecker-ci";
+    repo = "woodpecker";
+    rev = "v${version}";
+    sha256 = "sha256-HOOH3H2SXLcT2oW/xL80TO+ZSI+Haulnznpb4hlCQow=";
+  };
+
```
Comment (06kellyjac) [inline pkgs/development/tools/continuous-integration/woodpecker/common.nix:11]:
```suggestion
  yarnSha256 = "xyz";
```

**Review by 06kellyjac — COMMENTED** [review]
I'm not super familiar with passthru but yeah since it's an input to server you could replace it with anything, in this case an overriden frontend grabbed from passthru

so yeah that looks fine

---

As for the remaining eval/IFD issues I think it's because you need to set offlineCache or it just does it magically with IFD stuff

File: pkgs/development/tools/continuous-integration/woodpecker/agent.nix [hunk 3]
```diff
@@ -0,0 +1,23 @@
+{ lib, buildGoModule, fetchFromGitHub, mkYarnPackage }:
```
Comment (06kellyjac) [inline pkgs/development/tools/continuous-integration/woodpecker/agent.nix:1]:
```suggestion
{ lib, buildGoModule, fetchFromGitHub }:
```

Comment (ambroisie) [reply pkgs/development/tools/continuous-integration/woodpecker/agent.nix:1]: Ooops indeed.

**Comment by ambroisie** [comment]
Pushed some fixups (which will need to be rebased/squashed).

File: pkgs/development/tools/continuous-integration/woodpecker/common.nix [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+  yarnSha256 = "${lib.fakeSha256}";
```
Comment (ambroisie) [inline pkgs/development/tools/continuous-integration/woodpecker/common.nix:12]: Ooops, should be the actual checksum instead.

File: pkgs/development/tools/continuous-integration/woodpecker/frontend.nix [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+  packageJSON = ./woodpecker-package.json;
+  offlineCache = fetchYarnDeps {
+    yarnLock = "${src}/web/yarn.lock";
+    sha256 = yarnSha256;
+  };
+
+  buildPhase = ''
+    runHook preBuild
+
+    yarn build
```
[the remark this answers is not in the document]
Comment (ambroisie) [reply pkgs/development/tools/continuous-integration/woodpecker/frontend.nix:25]: Doesn't build.

**Comment by wyndon** [comment]
Result of `nixpkgs-review pr 178441` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>woodpecker-agent</li>
    <li>woodpecker-cli</li>
    <li>woodpecker-server</li>
  </ul>
</details>

**Review by 06kellyjac — APPROVED** [review]

File: pkgs/top-level/all-packages.nix [hunk 6]
```diff
@@ -11491,6 +11491,14 @@ with pkgs;
 
   woff2 = callPackage ../development/web/woff2 { };
 
+  woodpecker-agent = callPackage ../development/tools/continuous-integration/woodpecker/agent.nix { };
+
+  woodpecker-cli = callPackage ../development/tools/continuous-integration/woodpecker/cli.nix { };
+
+  woodpecker-server = callPackage ../development/tools/continuous-integration/woodpecker/server.nix {
+    woodpecker-frontend = callPackage ../development/tools/continuous-integration/woodpecker/frontend.nix { };
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:11499]: This should be available in passthru.

File: pkgs/development/tools/continuous-integration/woodpecker/update.sh [hunk 7]
```diff
@@ -0,0 +1,44 @@
+#!/usr/bin/env nix-shell
+#!nix-shell -i bash -p nix wget prefetch-yarn-deps nix-prefetch-github jq
+
+if [ "$#" -gt 1 ] || [[ "$1" == -* ]]; then
```
Comment (SuperSandro2000) [inline pkgs/development/tools/continuous-integration/woodpecker/update.sh:4]:
```suggestion
if [[ $# -gt 1 || $1 == -* ]]; then
```

File: pkgs/development/tools/continuous-integration/woodpecker/update.sh [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+set -x
+
+cd "$(dirname "$0")"
+version="$1"
+
+set -euo pipefail
+
+if [ -z "$version" ]; then
+    version="$(wget -O- "https://api.github.com/repos/woodpecker-ci/woodpecker/releases?per_page=1" | jq -r '.[0].tag_name')"
```
Comment (SuperSandro2000) [inline pkgs/development/tools/continuous-integration/woodpecker/update.sh:18]: Missing optional GitHub token.

File: pkgs/development/tools/continuous-integration/woodpecker/server.nix [hunk 9]
```diff
@@ -0,0 +1,33 @@
+{ lib, buildGoModule, fetchFromGitHub, woodpecker-frontend }:
+let
+  inherit (import ./common.nix { inherit lib fetchFromGitHub; })
```
Comment (SuperSandro2000) [inline pkgs/development/tools/continuous-integration/woodpecker/server.nix:3]: Use callPackage

File: pkgs/development/tools/continuous-integration/woodpecker/server.nix [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+    postBuild
+    ;
+in
+buildGoModule {
+  pname = "woodpecker-server";
+  inherit version src ldflags postBuild;
+  vendorSha256 = null;
+
+  postPatch = ''
+    cp -r ${woodpecker-frontend} web/dist
```
Comment (SuperSandro2000) [inline pkgs/development/tools/continuous-integration/woodpecker/server.nix:17]: Using a symlink avoids copying all the files.

File: pkgs/development/tools/continuous-integration/woodpecker/frontend.nix [hunk 11]
```diff
@@ -0,0 +1,45 @@
+{ lib, fetchFromGitHub, fetchYarnDeps, mkYarnPackage }:
+let
+  inherit (import ./common.nix { inherit lib fetchFromGitHub; })
```
Comment (SuperSandro2000) [inline pkgs/development/tools/continuous-integration/woodpecker/frontend.nix:3]: If we assign this to a variable and inherit from it we don't need to maintain two lists to inherit things.

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/development/tools/continuous-integration/woodpecker/frontend.nix (hunk 11 above)
Comment (06kellyjac) [reply pkgs/development/tools/continuous-integration/woodpecker/frontend.nix:3]: I guess you mean this:

```
inherit (import ./common.nix { }) a b c;

buildGoModule {
  # xyz
  inherit a b c;
}
```

to

```
common = inherit (import ./common.nix { });

buildGoModule {
  # xyz
  inherit (common) a b c;
}
```

---

I was also debating whether we should just have a general `commonGo` that holds stuff like `src` `version` `vendorSha256` `ldflags` etc and just do

```
inherit (import ./common.nix { }) commonGo meta;

buildGoModule (commonGo // {
  # xyz
  meta = meta // {
  };
})
```

File: pkgs/top-level/all-packages.nix (hunk 6 above)
Comment (ambroisie) [reply pkgs/top-level/all-packages.nix:11499]: I don't understand what you mean, it is in `passthru`.

**Pull request merged**